src/main/kotlin/net/maizegenetics/phgv2/cli/CreateFastaFromHvcf.kt

```diff
@@ -162,15 +173,47 @@ class CreateFastaFromHvcf : CliktCommand( help = "Create a fasta file from a hvc
             check(altHeaders.containsKey(hapId)) { "Haplotype ID $hapId not found in ALT Header" }
             val altMetaData = altHeaders[hapId]
             //Need to subtract 1 from start as it uses 0 based format
-            val range = "${altMetaData!!.contig}@${sampleName}:${altMetaData!!.start-1}-${altMetaData!!.end-1}"
-            val outputDisplayName = "${altMetaData!!.contig}:${altMetaData!!.start-1}-${altMetaData!!.end-1}"
-            Triple(range, outputDisplayName, HaplotypeSequence(hapId, "", altMetaData.refRange, it.contig, it.start, it.end, altMetaData!!.contig, altMetaData!!.start, altMetaData!!.end))
+            val regions =  altMetaData!!.regions
+            val queryRanges = mutableListOf<String>()
+            val displayRanges = mutableListOf<String>()
+            for(region in regions) {
+                queryRanges.add("${region.first.contig}@${sampleName}:${region.first.position-1}-${region.second.position-1}")
+                displayRanges.add("${region.first.contig}:${region.first.position-1}-${region.second.position-1}")
+            }
+            Triple(queryRanges, displayRanges, HaplotypeSequence(hapId, "", altMetaData.refRange, it.contig, it.start, it.end, regions))
         }
 
-        val ranges = rangesAndOtherInfo.map { it.first }
+        //Create a list of ranges we need to extract.
+        //Go through and split them into sets of non-identical ranges
+        //Pull all the non-identical ranges in a multithreaded fashion
+        //Store in Map<Range,Sequence>
+        val ranges = rangesAndOtherInfo.flatMap { it.first }
         val seqs = retrieveAgcContigs(dbPath,ranges)
 
-        return rangesAndOtherInfo.map { it.third.copy(sequence = seqs[it.second]!!.seq()) }
+        return rangesAndOtherInfo.map { it.third.copy(sequence = buildHapSeq(seqs, it.second,it.third)) }
+    }
+
+    /**
+     * Function to build the haplotype sequence based on the list of display regions and the given haplotype sequence object.
+     * The sequence is already extracted out of AGC and stored in the seqs map.
+     */
+    fun buildHapSeq(seqs: Map<String,NucSeq> ,displayRegions : List<String>, hapSeqObjects: HaplotypeSequence) : String {
+        val hapSeqRegions = hapSeqObjects.asmRegions
+
+        return displayRegions.mapIndexed{ idx, currentDisplayRegion ->
+            val currentHapSeqRegion = hapSeqRegions[idx]
+            val seq = seqs[currentDisplayRegion]!!
+
+            //Means it is the first region
+            if(currentHapSeqRegion.first.position > currentHapSeqRegion.second.position) {
+                //Means it needs to be reverse complemented
```

Comment:
Similar to Peter's comment above - can you add more commentary here on what "it" is?

Reply:
I updated the comment here and above it should be more clear now.